File: packages/retrieval-py/betterdb_retrieval/retriever.py
from __future__ import annotations

import json
import logging
import time
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Awaitable, Callable, Optional, Protocol, runtime_checkable

from betterdb_valkey_search_kit import (
    FtSearchHit,
    encode_float32,
    is_index_not_found_error,
    parse_dimension_from_info,
    parse_ft_info_stats,
    parse_ft_search_response,
)

from .discovery import (
    RETRIEVAL_CACHE_TYPE,
    RETRIEVAL_VERSION,
    REGISTRY_KEY,
    build_retrieval_marker,
)
from .fields import RESERVED_FIELD_NAMES, SCORE_FIELD, TEXT_FIELD
from .ft_create import (
    build_ft_create_args,
    index_name,
    key_prefix,
    resolve_vector_field_name,
)
from .ft_search import QueryFilter, build_ft_search_query
from .health import IndexHealthSnapshot, RecallEstimator, parse_percent_indexed
from .schema import FtCapabilities, RetrievalSchema
from .telemetry import RetrievalMetrics, RetrievalOperation, RetrievalTracer

_logger = logging.getLogger(__name__)

EmbedFn = Callable[[str], Awaitable[list[float]]]

# Atomic compare-and-set for the shared registry field. REGISTRY_KEY is keyed by
# name and shared with agent-cache, so a plain HGET -> compare -> HSET/HDEL has a
# TOCTOU window in which a foreign marker written in between gets clobbered. These
# scripts collapse read-compare-write into one server-side round trip.
_REGISTER_SCRIPT = """
local raw = redis.call('HGET', KEYS[1], ARGV[1])
if raw then
  local ok, parsed = pcall(cjson.decode, raw)
  if ok and type(parsed) == 'table' and parsed.type and parsed.type ~= ARGV[3] then
    return parsed.type
  end
end
redis.call('HSET', KEYS[1], ARGV[1], ARGV[2])
return false
"""

_UNREGISTER_SCRIPT = """
local raw = redis.call('HGET', KEYS[1], ARGV[1])
if raw then
  local ok, parsed = pcall(cjson.decode, raw)
  if ok and type(parsed) == 'table' and parsed.type == ARGV[2] then
    return redis.call('HDEL', KEYS[1], ARGV[1])
  end
end
return 0
"""


@dataclass
class QueryHit:
    id: str
    # Raw KNN ``__score`` from valkey-search: a vector **distance**, not a
    # similarity. Lower means closer (a perfect match approaches 0), so rank
    # ascending. Do not assume higher is better.
    score: float
    text: str
    fields: dict[str, str]


RerankFn = Callable[[str, list[QueryHit]], Awaitable[list[QueryHit]]]


@dataclass
class IndexDescription:
    name: str
    dims: int
    num_docs: int
    indexing_state: str


@dataclass
class UpsertEntry:
    id: str
    text: str
    fields: dict[str, "str | int | float"] = field(default_factory=dict)


@runtime_checkable
class RetrieverClient(Protocol):
    """A minimal async Valkey client surface (valkey-py is a drop-in)."""

    async def execute_command(self, *args: Any) -> Any: ...


def _is_positive_int(value: object) -> bool:
    if isinstance(value, bool):
        return False
    if isinstance(value, int):
        return value > 0
    if isinstance(value, float):
        return value.is_integer() and value > 0
    return False


class Retriever:
    def __init__(
        self,
        client: RetrieverClient,
        name: str,
        schema: RetrievalSchema,
        *,
        capabilities: FtCapabilities | None = None,
        embed_fn: EmbedFn | None = None,
        rerank_fn: RerankFn | None = None,
        recall_estimator: RecallEstimator | None = None,
        metrics: RetrievalMetrics | None = None,
        tracer: RetrievalTracer | None = None,
    ) -> None:
        self._client = client
        self._name = name
        self._schema = schema
        self._capabilities = capabilities
        self._embed_fn = embed_fn
        self._rerank_fn = rerank_fn
        self._recall_estimator = recall_estimator
        self._metrics = metrics
        self._tracer = tracer
        self._resolved_dims: Optional[int] = None

    async def _instrument(
        self, operation: RetrievalOperation, fn: Callable[[], Awaitable[Any]]
    ) -> Any:
        span = self._tracer.start_span(f"retrieval.{operation}") if self._tracer else None
        start = time.perf_counter()
        try:
            return await fn()
        finally:
            if self._metrics is not None:
                self._metrics.observe_operation(operation, time.perf_counter() - start)
            if span is not None:
                span.end()

    async def _resolve_dims(self) -> int:
        declared = self._schema["vector"].get("dims")
        if declared is not None:
            if not _is_positive_int(declared):
                raise ValueError(f"schema.vector.dims must be a positive integer, got: {declared}")
            return int(declared)
        if self._resolved_dims is not None:
            return self._resolved_dims
        if self._embed_fn is None:
            raise ValueError(
                "Cannot resolve vector dimension: provide schema.vector.dims or an embedFn"
            )
        probe = await self._embed_fn("probe")
        if self._metrics is not None:
            self._metrics.record_embedding_call()
        if len(probe) == 0:
            raise ValueError(
                "Cannot resolve vector dimension: embedFn returned a zero-length probe embedding"
            )
        self._resolved_dims = len(probe)
        return self._resolved_dims

    async def create_index(self) -> None:
        try:
            await self._client.execute_command("FT.INFO", index_name(self._name))
            return
        except Exception as err:
            if not is_index_not_found_error(err):
                raise
        dims = await self._resolve_dims()
        schema: RetrievalSchema = {
            "fields": self._schema["fields"],
            "vector": {**self._schema["vector"], "dims": dims},
        }
        try:
            await self._client.execute_command(
                "FT.CREATE", *build_ft_create_args(self._name, schema, self._capabilities)
            )
        except Exception as err:
            # Tolerate a concurrent creation: another worker may create the index
            # between our FT.INFO probe and this FT.CREATE (common on multi-worker
            # boot). The idempotent contract holds as long as the index exists.
            if "already exists" not in str(err).lower():
                raise

    def _assert_no_reserved_fields(self, entry: UpsertEntry, vector_field: str) -> None:
        for field_name in entry.fields:
            if field_name in RESERVED_FIELD_NAMES or field_name == vector_field:
                raise ValueError(
                    f"Entry '{entry.id}' uses reserved field name '{field_name}'; "
                    "choose a different field name"
                )

    async def _embed(self, text: str) -> list[float]:
        if self._embed_fn is None:
            raise ValueError("Cannot embed text: provide an embedFn")
        dims = await self._resolve_dims()
        vector = await self._embed_fn(text)
        if self._metrics is not None:
            self._metrics.record_embedding_call()
        if len(vector) != dims:
            raise ValueError(
                f"Embedding dimension mismatch: index expects {dims}, embedFn returned {len(vector)}"
            )
        return vector

    async def upsert(self, entries: list[UpsertEntry]) -> None:
        await self._instrument("upsert", lambda: self._upsert_entries(entries))

    async def _upsert_entries(self, entries: list[UpsertEntry]) -> None:
        vector_field = resolve_vector_field_name(self._schema["vector"])
        writes: list[tuple[str, list[Any]]] = []
        for entry in entries:
            self._assert_no_reserved_fields(entry, vector_field)
            vector = await self._embed(entry.text)
            args: list[Any] = []
            for field_name, value in entry.fields.items():
                args.extend([field_name, str(value)])
            args.extend([vector_field, encode_float32(vector)])
            args.extend([TEXT_FIELD, entry.text])
            writes.append((f"{key_prefix(self._name)}{entry.id}", args))
        for key, args in writes:
            await self._client.execute_command("HSET", key, *args)

    async def delete(self, ids: list[str]) -> None:
        if len(ids) == 0:
            return
        keys = [f"{key_prefix(self._name)}{entry_id}" for entry_id in ids]
        await self._client.execute_command("DEL", *keys)

    async def drop_index(self) -> None:
        try:
            await self._client.execute_command("FT.DROPINDEX", index_name(self._name))
        except Exception as err:
            if not is_index_not_found_error(err):
                raise

    async def describe_index(self) -> IndexDescription:
        info = await self._client.execute_command("FT.INFO", index_name(self._name))
        stats = parse_ft_info_stats(info)
        return IndexDescription(
            name=self._name,
            dims=parse_dimension_from_info(info),
            num_docs=stats.num_docs,
            indexing_state=stats.indexing_state,
        )

    def _known_dims(self) -> Optional[int]:
        declared = self._schema["vector"].get("dims")
        if declared is not None and _is_positive_int(declared):
            return int(declared)
        return self._resolved_dims

    async def _query_vector_dims(self) -> Optional[int]:
        known = self._known_dims()
        if known is not None:
            return known
        if self._embed_fn is None:
            return None
        return await self._resolve_dims()

    async def _resolve_query_vector(
        self, text: Optional[str], vector: Optional[list[float]]
    ) -> list[float]:
        if vector is not None and text is not None:
            raise ValueError("query accepts either text or a precomputed vector, not both")
        if vector is not None:
            dims = await self._query_vector_dims()
            if dims is not None and len(vector) != dims:
                raise ValueError(
                    f"Query vector dimension mismatch: index expects {dims}, got {len(vector)}"
                )
            return vector
        if text is not None:
            return await self._embed(text)
        raise ValueError("query requires either text or a precomputed vector")

    def _map_hit(self, hit: FtSearchHit) -> QueryHit:
        prefix = key_prefix(self._name)
        entry_id = hit["key"]
        if entry_id.startswith(prefix):
            entry_id = entry_id[len(prefix) :]
        vector_field = resolve_vector_field_name(self._schema["vector"])
        fields: dict[str, str] = {}
        for field_name, value in hit["fields"].items():
            if field_name in (TEXT_FIELD, SCORE_FIELD, vector_field):
                continue
            fields[field_name] = value
        return QueryHit(
            id=entry_id,
            score=float(hit["fields"][SCORE_FIELD]),
            text=hit["fields"].get(TEXT_FIELD, ""),
            fields=fields,
        )

    def _resolve_rerank(
        self, hybrid: Optional[str], text: Optional[str]
    ) -> Optional[tuple[RerankFn, str]]:
        if hybrid != "rerank":
            return None
        if self._rerank_fn is None:
            raise ValueError("query(hybrid='rerank') requires a rerankFn")
        if text is None:
            raise ValueError("query(hybrid='rerank') requires text to rerank against")
        return (self._rerank_fn, text)

    async def query(
        self,
        *,
        k: int,
        text: Optional[str] = None,
        vector: Optional[list[float]] = None,
        filter: Optional[QueryFilter] = None,
        hybrid: Optional[str] = None,
    ) -> list[QueryHit]:
        if not _is_positive_int(k):
            raise ValueError(f"query k must be a positive integer, got: {k}")
        rerank = self._resolve_rerank(hybrid, text)
        return await self._instrument(
            "query", lambda: self._run_query(text, vector, k, filter, rerank)
        )

    async def _run_query(
        self,
        text: Optional[str],
        vector: Optional[list[float]],
        k: int,
        filter: Optional[QueryFilter],
        rerank: Optional[tuple[RerankFn, str]],
    ) -> list[QueryHit]:
        resolved_vector = await self._resolve_query_vector(text, vector)
        query_string = build_ft_search_query(self._schema, k, filter)
        raw = await self._client.execute_command(
            "FT.SEARCH",
            index_name(self._name),
            query_string,
            "PARAMS",
            "2",
            "vec",
            encode_float32(resolved_vector),
            "LIMIT",
            "0",
            str(k),
            "DIALECT",
            "2",
        )
        hits = [self._map_hit(hit) for hit in parse_ft_search_response(raw)]
        result = hits
        if rerank is not None:
            rerank_fn, rerank_text = rerank
            result = await rerank_fn(rerank_text, hits)
        if self._metrics is not None:
            self._metrics.record_query_results(len(result))
        return result

    async def register(self) -> None:
        # The registry field is keyed by name and shared with agent-cache. Compare
        # the existing marker's type and write ours in a single atomic round trip
        # (_REGISTER_SCRIPT) so a foreign marker can't be clobbered through a
        # check-then-act window. The script returns the foreign type when it skips.
        marker = build_retrieval_marker(
            name=self._name,
            version=RETRIEVAL_VERSION,
            started_at=datetime.now(timezone.utc).isoformat(),
        )
        foreign = await self._client.execute_command(
            "EVAL",
            _REGISTER_SCRIPT,
            1,
            REGISTRY_KEY,
            self._name,
            json.dumps(marker),
            RETRIEVAL_CACHE_TYPE,
        )
        if foreign is not None:
            if isinstance(foreign, bytes):
                foreign = foreign.decode()
            _logger.warning(
                "retrieval discovery: registry field '%s' already holds a '%s' marker; "
                "skipping registration",
                self._name,
                foreign,
            )

    async def unregister(self) -> None:
        # Only delete a marker we own — compared and HDEL'd in one atomic round
        # trip (_UNREGISTER_SCRIPT) so we never delete a foreign cache type's field.
        await self._client.execute_command(
            "EVAL",
            _UNREGISTER_SCRIPT,
            1,
            REGISTRY_KEY,
            self._name,
            RETRIEVAL_CACHE_TYPE,
        )

    async def health(self) -> IndexHealthSnapshot:
        info = await self._client.execute_command("FT.INFO", index_name(self._name))
        stats = parse_ft_info_stats(info)
        snapshot = IndexHealthSnapshot(
            name=self._name,
            num_docs=stats.num_docs,
            indexing_state=stats.indexing_state,
            dims=parse_dimension_from_info(info),
            percent_indexed=parse_percent_indexed(info),
            estimated_recall=None,
        )
        if self._recall_estimator is not None:
            snapshot.estimated_recall = self._recall_estimator(snapshot)
        return snapshot

File: packages/retrieval/src/retriever.ts
import {
  encodeFloat32,
  isIndexNotFoundError,
  parseDimensionFromInfo,
  parseFtInfoStats,
  parseFtSearchResponse,
  type FtSearchHit,
} from '@betterdb/valkey-search-kit';
import type { RetrievalSchema, FtCapabilities } from './schema';
import { buildFtCreateArgs, indexName, keyPrefix, resolveVectorFieldName } from './ft-create';
import { buildFtSearchQuery, type QueryFilter } from './ft-search';
import { TEXT_FIELD, SCORE_FIELD, RESERVED_FIELD_NAMES } from './fields';
import {
  buildRetrievalMarker,
  REGISTRY_KEY,
  RETRIEVAL_CACHE_TYPE,
  RETRIEVAL_VERSION,
} from './discovery';
import { parsePercentIndexed, type IndexHealthSnapshot, type RecallEstimator } from './health';
import type { RetrievalMetrics, RetrievalTracer, RetrievalOperation } from './telemetry';
import { createAnalytics, NOOP_ANALYTICS, type Analytics, type AnalyticsOptions } from './analytics';

// Atomic compare-and-set for the shared registry field. REGISTRY_KEY is keyed by
// name and shared with agent-cache, so a plain HGET -> compare -> HSET/HDEL has a
// TOCTOU window in which a foreign marker written in between gets clobbered. These
// scripts collapse read-compare-write into one server-side round trip.
const REGISTER_SCRIPT = `
local raw = redis.call('HGET', KEYS[1], ARGV[1])
if raw then
  local ok, parsed = pcall(cjson.decode, raw)
  if ok and type(parsed) == 'table' and parsed.type and parsed.type ~= ARGV[3] then
    return parsed.type
  end
end
redis.call('HSET', KEYS[1], ARGV[1], ARGV[2])
return false
`;

const UNREGISTER_SCRIPT = `
local raw = redis.call('HGET', KEYS[1], ARGV[1])
if raw then
  local ok, parsed = pcall(cjson.decode, raw)
  if ok and type(parsed) == 'table' and parsed.type == ARGV[2] then
    return redis.call('HDEL', KEYS[1], ARGV[1])
  end
end
return 0
`;

export type EmbedFn = (text: string) => Promise<number[]>;

export type RerankFn = (queryText: string, hits: QueryHit[]) => Promise<QueryHit[]>;

export interface QueryHit {
  id: string;
  /**
   * Raw KNN `__score` from valkey-search: a vector **distance**, not a
   * similarity. Lower means closer (a perfect match approaches 0), so rank
   * ascending. Do not assume higher is better.
   */
  score: number;
  text: string;
  fields: Record<string, string>;
}

export interface QueryOptions {
  text?: string;
  vector?: number[];
  k: number;
  filter?: QueryFilter;
  hybrid?: 'rerank';
}

export interface RetrieverClient {
  call(command: string, ...args: (string | Buffer | number)[]): Promise<unknown>;
}

export interface IndexDescription {
  name: string;
  dims: number;
  numDocs: number;
  indexingState: string;
}

export interface RetrieverOptions {
  client: RetrieverClient;
  name: string;
  schema: RetrievalSchema;
  capabilities?: FtCapabilities;
  embedFn?: EmbedFn;
  rerankFn?: RerankFn;
  recallEstimator?: RecallEstimator;
  metrics?: RetrievalMetrics;
  tracer?: RetrievalTracer;
  analytics?: AnalyticsOptions;
}

export interface UpsertEntry {
  id: string;
  text: string;
  fields: Record<string, string | number>;
}

export class Retriever {
  private readonly client: RetrieverClient;
  private readonly name: string;
  private readonly schema: RetrievalSchema;
  private readonly capabilities?: FtCapabilities;
  private readonly embedFn?: EmbedFn;
  private readonly rerankFn?: RerankFn;
  private readonly recallEstimator?: RecallEstimator;
  private readonly metrics?: RetrievalMetrics;
  private readonly tracer?: RetrievalTracer;
  private resolvedDims?: number;
  private readonly analyticsOptions?: AnalyticsOptions;
  private analytics: Analytics = NOOP_ANALYTICS;
  private analyticsStarted = false;

  constructor(options: RetrieverOptions) {
    this.client = options.client;
    this.name = options.name;
    this.schema = options.schema;
    this.capabilities = options.capabilities;
    this.embedFn = options.embedFn;
    this.rerankFn = options.rerankFn;
    this.recallEstimator = options.recallEstimator;
    this.metrics = options.metrics;
    this.tracer = options.tracer;
    this.analyticsOptions = options.analytics;
  }

  // Fire-once: defer analytics startup to the first index-lifecycle call so the
  // real client is awaited before any event is captured (the constructor cannot
  // await). Never lets analytics break the retriever.
  private async ensureAnalyticsStarted(): Promise<void> {
    if (this.analyticsStarted) {
      return;
    }
    this.analyticsStarted = true;
    try {
      const analytics = await createAnalytics({
        apiKey: this.analyticsOptions?.apiKey,
        host: this.analyticsOptions?.host,
        disabled: this.analyticsOptions?.disabled,
      });
      this.analytics = analytics;
      await analytics.init(this.client, this.name, {
        fieldCount: this.schema.fields.length,
        vectorMetric: this.schema.vector.metric,
        vectorAlgorithm: this.schema.vector.algorithm,
        hasEmbedFn: this.embedFn !== undefined,
        hasRerankFn: this.rerankFn !== undefined,
      });
    } catch {
      this.analytics = NOOP_ANALYTICS;
    }
  }

  /** Tear down product analytics (flushes any pending events). */
  async close(): Promise<void> {
    await this.analytics.shutdown();
  }

  private async instrument<T>(operation: RetrievalOperation, fn: () => Promise<T>): Promise<T> {
    const span = this.tracer?.startSpan(`retrieval.${operation}`);
    const start = performance.now();
    try {
      return await fn();
    } finally {
      this.metrics?.observeOperation(operation, (performance.now() - start) / 1000);
      span?.end();
    }
  }

  private async resolveDims(): Promise<number> {
    const declared = this.schema.vector.dims;
    if (declared !== undefined) {
      if (!Number.isInteger(declared) || declared <= 0) {
        throw new Error(`schema.vector.dims must be a positive integer, got: ${declared}`);
      }
      return declared;
    }
    if (this.resolvedDims !== undefined) {
      return this.resolvedDims;
    }
    if (this.embedFn === undefined) {
      throw new Error('Cannot resolve vector dimension: provide schema.vector.dims or an embedFn');
    }
    const probe = await this.embedFn('probe');
    this.metrics?.recordEmbeddingCall();
    if (probe.length === 0) {
      throw new Error(
        'Cannot resolve vector dimension: embedFn returned a zero-length probe embedding',
      );
    }
    this.resolvedDims = probe.length;
    return this.resolvedDims;
  }

  async createIndex(): Promise<void> {
    await this.ensureAnalyticsStarted();
    try {
      await this.client.call('FT.INFO', indexName(this.name));
      return;
    } catch (err) {
      if (!isIndexNotFoundError(err)) {
        throw err;
      }
    }
    const dims = await this.resolveDims();
    const schema: RetrievalSchema = {
      ...this.schema,
      vector: { ...this.schema.vector, dims },
    };
    try {
      await this.client.call(
        'FT.CREATE',
        ...buildFtCreateArgs(this.name, schema, this.capabilities),
      );
    } catch (err) {
      // Tolerate a concurrent creation: another worker may create the index
      // between our FT.INFO probe and this FT.CREATE (common on multi-worker
      // boot). The idempotent contract holds as long as the index exists.
      if (!String(err).toLowerCase().includes('already exists')) {
        throw err;
      }
      // This worker did not create the index — skip the telemetry event so a
      // multi-worker boot does not over-count index creation.
      return;
    }
    this.analytics.capture('index_created', { dims });
  }

  private assertNoReservedFields(entry: UpsertEntry, vectorField: string): void {
    for (const field of Object.keys(entry.fields)) {
      if (RESERVED_FIELD_NAMES.includes(field) || field === vectorField) {
        throw new Error(
          `Entry '${entry.id}' uses reserved field name '${field}'; choose a different field name`,
        );
      }
    }
  }

  private async embed(text: string): Promise<number[]> {
    if (this.embedFn === undefined) {
      throw new Error('Cannot embed text: provide an embedFn');
    }
    const dims = await this.resolveDims();
    const vector = await this.embedFn(text);
    this.metrics?.recordEmbeddingCall();
    if (vector.length !== dims) {
      throw new Error(
        `Embedding dimension mismatch: index expects ${dims}, embedFn returned ${vector.length}`,
      );
    }
    return vector;
  }

  async upsert(entries: UpsertEntry[]): Promise<void> {
    return this.instrument('upsert', () => this.upsertEntries(entries));
  }

  private async upsertEntries(entries: UpsertEntry[]): Promise<void> {
    const vectorField = resolveVectorFieldName(this.schema.vector);
    const writes: { key: string; args: (string | Buffer)[] }[] = [];
    for (const entry of entries) {
      this.assertNoReservedFields(entry, vectorField);
      const vector = await this.embed(entry.text);
      const args: (string | Buffer)[] = [];
      for (const [field, value] of Object.entries(entry.fields)) {
        args.push(field, String(value));
      }
      args.push(vectorField, encodeFloat32(vector));
      args.push(TEXT_FIELD, entry.text);
      writes.push({ key: `${keyPrefix(this.name)}${entry.id}`, args });
    }
    for (const write of writes) {
      await this.client.call('HSET', write.key, ...write.args);
    }
  }

  async delete(ids: string[]): Promise<void> {
    if (ids.length === 0) {
      return;
    }
    const keys = ids.map((id) => `${keyPrefix(this.name)}${id}`);
    await this.client.call('DEL', ...keys);
  }

  async dropIndex(): Promise<void> {
    try {
      await this.client.call('FT.DROPINDEX', indexName(this.name));
    } catch (err) {
      if (!isIndexNotFoundError(err)) {
        throw err;
      }
    }
  }

  async describeIndex(): Promise<IndexDescription> {
    const info = (await this.client.call('FT.INFO', indexName(this.name))) as unknown[];
    const stats = parseFtInfoStats(info);
    return {
      name: this.name,
      dims: parseDimensionFromInfo(info),
      numDocs: stats.numDocs,
      indexingState: stats.indexingState,
    };
  }

  private knownDims(): number | undefined {
    const declared = this.schema.vector.dims;
    if (declared !== undefined && Number.isInteger(declared) && declared > 0) {
      return declared;
    }
    return this.resolvedDims;
  }

  private async queryVectorDims(): Promise<number | undefined> {
    const known = this.knownDims();
    if (known !== undefined) {
      return known;
    }
    if (this.embedFn === undefined) {
      return undefined;
    }
    return this.resolveDims();
  }

  private async resolveQueryVector(options: QueryOptions): Promise<number[]> {
    if (options.vector !== undefined && options.text !== undefined) {
      throw new Error('query accepts either text or a precomputed vector, not both');
    }
    if (options.vector !== undefined) {
      const dims = await this.queryVectorDims();
      if (dims !== undefined && options.vector.length !== dims) {
        throw new Error(
          `Query vector dimension mismatch: index expects ${dims}, got ${options.vector.length}`,
        );
      }
      return options.vector;
    }
    if (options.text !== undefined) {
      return this.embed(options.text);
    }
    throw new Error('query requires either text or a precomputed vector');
  }

  private mapHit(hit: FtSearchHit): QueryHit {
    const prefix = keyPrefix(this.name);
    let id = hit.key;
    if (hit.key.startsWith(prefix)) {
      id = hit.key.slice(prefix.length);
    }
    const vectorField = resolveVectorFieldName(this.schema.vector);
    const fields: Record<string, string> = {};
    for (const [field, value] of Object.entries(hit.fields)) {
      if (field === TEXT_FIELD || field === SCORE_FIELD || field === vectorField) {
        continue;
      }
      fields[field] = value;
    }
    return {
      id,
      score: Number(hit.fields[SCORE_FIELD]),
      text: hit.fields[TEXT_FIELD] ?? '',
      fields,
    };
  }

  private resolveRerank(options: QueryOptions): { fn: RerankFn; text: string } | null {
    if (options.hybrid !== 'rerank') {
      return null;
    }
    if (this.rerankFn === undefined) {
      throw new Error("query({ hybrid: 'rerank' }) requires a rerankFn");
    }
    if (options.text === undefined) {
      throw new Error("query({ hybrid: 'rerank' }) requires text to rerank against");
    }
    return { fn: this.rerankFn, text: options.text };
  }

  async query(options: QueryOptions): Promise<QueryHit[]> {
    if (!Number.isInteger(options.k) || options.k <= 0) {
      throw new Error(`query k must be a positive integer, got: ${options.k}`);
    }
    const rerank = this.resolveRerank(options);
    return this.instrument('query', () => this.runQuery(options, rerank));
  }

  private async runQuery(
    options: QueryOptions,
    rerank: { fn: RerankFn; text: string } | null,
  ): Promise<QueryHit[]> {
    const vector = await this.resolveQueryVector(options);
    const queryString = buildFtSearchQuery(this.schema, options.k, options.filter);
    const raw = await this.client.call(
      'FT.SEARCH',
      indexName(this.name),
      queryString,
      'PARAMS',
      '2',
      'vec',
      encodeFloat32(vector),
      'LIMIT',
      '0',
      String(options.k),
      'DIALECT',
      '2',
    );
    const hits = parseFtSearchResponse(raw).map((hit) => this.mapHit(hit));
    let result = hits;
    if (rerank !== null) {
      result = await rerank.fn(rerank.text, hits);
    }
    this.metrics?.recordQueryResults(result.length);
    return result;
  }

  async register(): Promise<void> {
    // The registry field is keyed by name and shared with agent-cache. Compare
    // the existing marker's type and write ours in a single atomic round trip
    // (REGISTER_SCRIPT) so a foreign marker can't be clobbered through a
    // check-then-act window. The script returns the foreign type when it skips.
    const marker = buildRetrievalMarker({
      name: this.name,
      version: RETRIEVAL_VERSION,
      startedAt: new Date().toISOString(),
    });
    const foreign = await this.client.call(
      'EVAL',
      REGISTER_SCRIPT,
      1,
      REGISTRY_KEY,
      this.name,
      JSON.stringify(marker),
      RETRIEVAL_CACHE_TYPE,
    );
    if (foreign !== null && foreign !== undefined) {
      console.warn(
        `retrieval discovery: registry field '${this.name}' already holds a '${String(foreign)}' marker; skipping registration`,
      );
    }
  }

  async unregister(): Promise<void> {
    // Only delete a marker we own — compared and HDEL'd in one atomic round trip
    // (UNREGISTER_SCRIPT) so we never delete a foreign cache type's field.
    await this.client.call(
      'EVAL',
      UNREGISTER_SCRIPT,
      1,
      REGISTRY_KEY,
      this.name,
      RETRIEVAL_CACHE_TYPE,
    );
  }

  async health(): Promise<IndexHealthSnapshot> {
    const info = (await this.client.call('FT.INFO', indexName(this.name))) as unknown[];
    const stats = parseFtInfoStats(info);
    const snapshot = {
      name: this.name,
      numDocs: stats.numDocs,
      indexingState: stats.indexingState,
      dims: parseDimensionFromInfo(info),
      percentIndexed: parsePercentIndexed(info),
    };
    let estimatedRecall: number | null = null;
    if (this.recallEstimator !== undefined) {
      estimatedRecall = this.recallEstimator(snapshot);
    }
    return { ...snapshot, estimatedRecall };
  }
}
